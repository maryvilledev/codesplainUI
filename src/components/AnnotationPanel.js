import React, { PropTypes } from 'react';

import AnnotationDisplay from './AnnotationDisplay';
import AnnotationEditor from './AnnotationEditor';
import LineSnippet from './LineSnippet';

class AnnotationPanel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditing: props.annotation.length === 0,
    };
    this.handleSaveAnnotation = this.handleSaveAnnotation.bind(this);
    this.toggleEditState = this.toggleEditState.bind(this);
  }

  toggleEditState() {
    this.setState({
      isEditing: !this.state.isEditing,
    });
  }

  handleSaveAnnotation(annotation) {
    this.props.saveAnnotation(annotation);
    this.toggleEditState();
  }

  render() {
    const {
      annotation,
      closeAnnotation,
      snippetInformation: {
        lineNumber,
        lineText,
      }
    } = this.props;
    const { isEditing } = this.state;
    const shouldRenderEditor = annotation.length === 0 || isEditing;
    return (
      <div>
        <LineSnippet
          lineNumber={lineNumber + 1}
          value={lineText}
        />
        { shouldRenderEditor ?
          <AnnotationEditor
            annotation={annotation}
            closeAnnotation={closeAnnotation}
            saveAnnotation={this.handleSaveAnnotation}
          /> :
          <AnnotationDisplay
            annotation={annotation}
            closeAnnotation={closeAnnotation}
            editAnnotation={this.toggleEditState}
          />
        }
      </div>
    );
  }

}

AnnotationPanel.propTypes = {
  annotation: PropTypes.string.isRequired,
  saveAnnotation: PropTypes.func.isRequired,
  closeAnnotation: PropTypes.func.isRequired,
  snippetInformation: PropTypes.shape({
    lineNumber: PropTypes.number.isRequired,
    lineText: PropTypes.string.isRequired,
  }).isRequired,
};

export default AnnotationPanel;
