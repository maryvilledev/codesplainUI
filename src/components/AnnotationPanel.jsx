import React, { PropTypes } from 'react';

import AnnotationDisplay from './AnnotationDisplay';
import AnnotationEditor from './AnnotationEditor';
import LineSnippet from './LineSnippet';

const style = {
  container: {
    padding: '3%',
  },
};

class AnnotationPanel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditing: props.annotation.length === 0,
    };
    this.handleSaveAnnotation = this.handleSaveAnnotation.bind(this);
    this.toggleEditState = this.toggleEditState.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    const {
      lineAnnotated: {
        lineNumber,
      },
    } = this.props;
    if (lineNumber !== nextProps.lineAnnotated.lineNumber) {
      this.setState({
        isEditing: nextProps.annotation.length === 0,
      });
    }
  }

  handleSaveAnnotation(annotation) {
    this.props.saveAnnotation(annotation);
    this.toggleEditState();
  }

  toggleEditState() {
    this.setState({
      isEditing: !this.state.isEditing,
    });
  }

  render() {
    const {
      annotation,
      closeAnnotation,
      getNextAnnotation,
      getPreviousAnnotation,
      hasNextAnnotation,
      hasPrevAnnotation,
      lineAnnotated: {
        lineNumber,
        lineText,
      },
    } = this.props;
    const { isEditing } = this.state;
    return (
      <div style={style.container}>
        <LineSnippet
          lineNumber={lineNumber + 1}
          value={lineText}
        />
        { isEditing ?
          <AnnotationEditor
            annotation={annotation}
            closeAnnotation={closeAnnotation}
            saveAnnotation={this.handleSaveAnnotation}
          /> :
          <AnnotationDisplay
            annotation={annotation}
            closeAnnotation={closeAnnotation}
            editAnnotation={this.toggleEditState}
            getNextAnnotation={getNextAnnotation}
            getPreviousAnnotation={getPreviousAnnotation}
            hasPrevAnnotation={hasPrevAnnotation}
            hasNextAnnotation={hasNextAnnotation}
          />
        }
      </div>
    );
  }
}

AnnotationPanel.propTypes = {
  annotation: PropTypes.string.isRequired,
  closeAnnotation: PropTypes.func.isRequired,
  saveAnnotation: PropTypes.func.isRequired,
  getNextAnnotation: PropTypes.func.isRequired,
  getPreviousAnnotation: PropTypes.func.isRequired,
  hasNextAnnotation: PropTypes.bool.isRequired,
  hasPrevAnnotation: PropTypes.bool.isRequired,
  lineAnnotated: PropTypes.shape({
    lineNumber: PropTypes.number,
    lineText: PropTypes.string,
  }).isRequired,
};

export default AnnotationPanel;
