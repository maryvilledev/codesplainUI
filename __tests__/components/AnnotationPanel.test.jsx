import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import AnnotationPanel from '../../src/components/AnnotationPanel';

const mockFunctionProps = {
  closeAnnotation: jest.fn(),
  getNextAnnotation: jest.fn(),
  getPreviousAnnotation: jest.fn(),
  saveAnnotation: jest.fn(),
};
const defaultProps = {
  ...mockFunctionProps,
  annotation: '',
  lineAnnotated: {
    lineNumber: 0,
    lineText: 'You don\'t know me!',
  },
  hasNextAnnotation: false,
  hasPrevAnnotation: false,
};

describe('<AnnotationPanel />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = node => shallow(node, { context: { muiTheme } });

  describe('snapshot tests', () => {
    it('matches snapshot when no annotation is saved for a line', () => {
      const wrapper = shallowWithContext(
        <AnnotationPanel
          {...defaultProps}
        />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('matches snapshot when an annotation is saved for a line', () => {
      const wrapper = shallowWithContext(
        <AnnotationPanel
          {...defaultProps}
          annotation="Wubba lubba dub dub!"
        />,
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('prop: lineAnnotated', () => {
    it('is forwarded to LineSnippet', () => {
      const lineAnnotated = {
        lineNumber: 0,
        lineText: 'You pass butter',
      };
      const wrapper = shallowWithContext(
        <AnnotationPanel
          {...defaultProps}
          annotation="some annotation"
          lineAnnotated={lineAnnotated}
        />,
      );
      const lineSnippet = wrapper.find('LineSnippet');
      expect(lineSnippet.prop('lineNumber')).toEqual(Number(lineAnnotated.lineNumber) + 1);
      expect(lineSnippet.prop('value')).toEqual(lineAnnotated.lineText);
    });
  });
});
