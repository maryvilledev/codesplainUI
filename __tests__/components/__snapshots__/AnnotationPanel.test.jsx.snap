exports[`<AnnotationPanel /> snapshot tests matches snapshot when an annotation is saved for a line 1`] = `
<div
  style={
    Object {
      "padding": "3%",
    }
  }>
  <LineSnippet
    lineNumber={1}
    value="You don\'t know me!" />
  <AnnotationDisplay
    annotation="Wubba lubba dub dub!"
    closeAnnotation={[Function]}
    editAnnotation={[Function]} />
</div>
`;

exports[`<AnnotationPanel /> snapshot tests matches snapshot when no annotation is saved for a line 1`] = `
<div
  style={
    Object {
      "padding": "3%",
    }
  }>
  <LineSnippet
    lineNumber={1}
    value="You don\'t know me!" />
  <AnnotationEditor
    annotation=""
    closeAnnotation={[Function]}
    saveAnnotation={[Function]} />
</div>
`;
